client/cli/src/config.rs

```diff
@@ -226,16 +229,51 @@ pub trait CliConfiguration<DCV: DefaultConfigurationValues = ()>: Sized {
 		base_path: &PathBuf,
 		cache_size: usize,
 		database: Database,
+		role: &Role
 	) -> Result<DatabaseConfig> {
-		Ok(match database {
+		let role_suffix = match role {
+			Role::Light => "light",
+			Role::Full | Role::Authority => "full",
+		};
+
+		let db_conf = match database {
 			Database::RocksDb => DatabaseConfig::RocksDb {
-				path: base_path.join("db"),
+				path: base_path.join("db").join(role_suffix),
 				cache_size,
 			},
 			Database::ParityDb => DatabaseConfig::ParityDb {
-				path: base_path.join("paritydb"),
+				path: base_path.join("paritydb").join(role_suffix),
 			},
-		})
+		};
+
+		if let Some(p) = db_conf.path() {
```

Comment:
It looks like the following migration code belongs in `sc-client` or `sc-client-db`. Right before opening the database. Cli should not be concerned with database migration. It would also remove the dependency on `sc-client-db`.

Reply:
Sure, this was the first thing I tried, but I actually struggled at this problem a bit, because the migration code needs to know the `Role` in which the client was started and the `DatabaseType` to be sure not to move the db to the wrong directory.

When I tried to inject the `Role` from the Cli to the client I had a hard time with the compiler. That's why I chose it the other way around. But maybe moving the `Role` enum to sc-client would work.

Reply:
Role is passed down as `db_type` here https://github.com/paritytech/substrate/blob/1d5abf01abafdb6c15bcd0172f5de09fd87c5fbf/client/db/src/utils.rs#L208

Reply:
Not really, there's two distinct (and overlapping) terms.

There's `Role` from https://github.com/paritytech/substrate/blob/master/client/network/src/config.rs#L148
and `DatabaseType` from https://github.com/paritytech/substrate/blob/master/client/db/src/utils.rs#L91

Role is coming from the cli as parameter and DatabaseType is determined from the database.

Reply:
As far as understand, `db_type` that's passed to `open_database` is derived from the CLI argument. The cli argument selects type of backend to use. Light backend opens the database as light here: https://github.com/paritytech/substrate/blob/master/client/db/src/light.rs#L76 
And full backend opens it as full here: https://github.com/paritytech/substrate/blob/master/client/db/src/lib.rs#L1082

`open_database` does indeed check that expected `db_type` matches the one in the database. But you can still use expected `db_type` in the same way you use role w.r.t. migration.

Reply:
If you check the database type before migration there should be no mismatch. 

I'm suggesting moving your migration code inside the `open_database` function.
Basically move all of the code from `open_database` to a helper function, e.g. `open_database_at`
So that the new version of `open_database` would do something like this:
1. Check the base path for an existing database. Try to open it with `open_database_at`, which will validate expected db_type. If it succeeds, close the database and do the migration.
2. Append "full" or "light" to base path. Open it with `open_database_at` and return

A bit of refactoring might be required to extract the meat of `open_database` to a helper function that does the heavy lifting of actually opening the database.